west/cmd/git.py

```diff
@@ -0,0 +1,255 @@
+# Copyright (c) 2018 Open Source Foundries Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''west git commands'''
+
+import argparse
+import collections
+import os
+import subprocess
+import sys
+
+import pykwalify.core
+import yaml
+
+from . import WestCommand
+from .. import log
+
+
+class Sync(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'sync',
+            'Clone/update all Git repositories specified in the manifest file')
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        projects = _all_projects(args)
+
+        # TODO: Error checking (if _git(...).returncode != 0: ...)
+
+        for project in projects:
```

Comment:
Needs a TODO to handle projects that have been removed as well.

(I guess this means persisting a representation of the current projects somewhere and comparing it to the updated manifest that `west sync` pulls down.)

Reply:
Projects removed from manifests should be removed from working copies. Otherwise there's old cruft around which no longer matters.

What to do with any local branches, though? Perhaps repo keeps git repositories and objects under `.repo`, and symlinks to those from working trees, precisely to handle this case without deleting local work. No symlinks on Windows, though.

Maybe error out and ask for a re-run with a `--force` option if there are repositories with local branches that would be removed, as a starting point?

Reply:
I'd also be mad if it removed any non-checked-in files I happened to have in there. "Lemme remove this stuff you no longer need for you" is a bit dangerous...

Personally, I'd be fine if they were just left behind, but we'll see.